cmds/core/shutdown/shutdown.go

```diff
@@ -33,21 +41,47 @@ var (
 		"-s":      unix.LINUX_REBOOT_CMD_SW_SUSPEND,
 	}
 	reboot = unix.Reboot
+	delay  = time.Sleep
 )
 
 func usage() {
 	log.Fatalf("shutdown [-h|-r|-s|halt|reboot|suspend] (defaults to halt)")
 }
 
 func main() {
-	if len(os.Args) == 1 {
-		os.Args = append(os.Args, "halt")
+	a := os.Args
+	if len(a) == 1 {
+		a = append(a, "halt")
 	}
-	op, ok := opcodes[os.Args[1]]
-	if !ok || len(os.Args) > 2 {
+	op, ok := opcodes[a[1]]
+	if !ok {
 		usage()
 	}
+	if len(a) < 3 {
+		a = append(a, "now")
+	}
+	when := time.Now()
+	switch {
+	case a[2] == "now":
+	case a[2][0] == '+':
+		m, err := time.ParseDuration(a[2][1:] + "m")
+		if err != nil {
+			log.Fatal(err)
+		}
+		when = when.Add(m)
+	default:
+		t, err := time.Parse(time.RFC3339, a[2])
+		if err != nil {
+			log.Fatal(err)
+		}
+		when = t
+	}
+
+	delay(when.Sub(time.Now()))
```

Comment:
Couldn't this be negative since when was `time.Now()` from a few moments ago.

Reply:
yes, and that's explicitly allowed. It means "return right now"